### PR TITLE
cmd/setup-etcd-environment: use all unicast address as candidate when discovering self

### PR DESCRIPTION
**Pull request by abhinavdahiya**
**- What I did**
the InterfaceAddrs [1] returns list of all unicast addreses from the system

[ipaddrs.go]
```go
package main

import (
	"fmt"
	"net"
)

func main() {
	addrs, err := ipAddrs()
	fmt.Println(addrs, err)
}

func ipAddrs() ([]string, error) {
	var ips []string
	addrs, err := net.InterfaceAddrs()
	if err != nil {
		return ips, err
	}
	for _, addr := range addrs {
		var ip net.IP
		switch v := addr.(type) {
		case *net.IPNet:
			ip = v.IP
		case *net.IPAddr:
			ip = v.IP
		}
		if ip == nil {
			continue
		}
		ip = ip.To4()
		if ip == nil {
			continue // not an ipv4 address
		}
		if !ip.IsGlobalUnicast() {
			continue // we only want global unicast address
		}
		ips = append(ips, ip.String())
	}

	return ips, nil
}
```
The new `ipAddrs` function returns all the ipv4 addresses rather than return one ip address for a defined interface
```console
$ ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: enp0s31f6: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc fq_codel state DOWN group default qlen 1000
    link/ether 8c:16:45:75:4c:dc brd ff:ff:ff:ff:ff:ff
3: wlp61s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 5c:5f:67:f4:d4:9f brd ff:ff:ff:ff:ff:ff
    inet 10.7.28.215/22 brd 10.7.31.255 scope global dynamic noprefixroute wlp61s0
       valid_lft 79238sec preferred_lft 79238sec
    inet6 fe80::8ef5:bf00:55ce:6e7d/64 scope link noprefixroute
       valid_lft forever preferred_lft forever
4: virbr0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default qlen 1000
    link/ether 52:54:00:fe:d6:b2 brd ff:ff:ff:ff:ff:ff
    inet 192.168.122.1/24 brd 192.168.122.255 scope global virbr0
       valid_lft forever preferred_lft forever
5: virbr0-nic: <BROADCAST,MULTICAST> mtu 1500 qdisc fq_codel master virbr0 state DOWN group default qlen 1000
    link/ether 52:54:00:fe:d6:b2 brd ff:ff:ff:ff:ff:ff
6: docker0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default
    link/ether 02:42:44:37:11:f8 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.1/16 brd 172.17.255.255 scope global docker0
       valid_lft forever preferred_lft forever
    inet6 fe80::42:44ff:fe37:11f8/64 scope link
       valid_lft forever preferred_lft forever
8: veth5fa51f8@if7: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master docker0 state UP group default
    link/ether e2:0a:07:18:fb:a4 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet6 fe80::e00a:7ff:fe18:fba4/64 scope link
       valid_lft forever preferred_lft forever
```
```console
go run test.go
[10.7.28.215 192.168.122.1 172.17.0.1] <nil>
```

The previous behavior of using the ip address from `eth0` as the only candidate for etcd's ip address does not work on bare metal machines.

for example, a bare metal machine on packet has following interfaces
```console
$ ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: enp0s20f0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 0c:c4:7a:e5:43:a2 brd ff:ff:ff:ff:ff:ff
    inet 147.75.69.215/31 brd 147.75.69.215 scope global dynamic noprefixroute enp0s20f0
       valid_lft 1253sec preferred_lft 1253sec
    inet6 fe80::c50a:7dda:dc01:1c24/64 scope link noprefixroute
       valid_lft forever preferred_lft forever
3: enp0s20f1: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq state DOWN group default qlen 1000
    link/ether 0c:c4:7a:e5:43:a3 brd ff:ff:ff:ff:ff:ff
4: cni0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default qlen 1000
    link/ether f2:87:f1:f4:28:56 brd ff:ff:ff:ff:ff:ff
    inet 10.88.0.1/16 scope global cni0
       valid_lft forever preferred_lft forever
    inet6 fe80::f087:f1ff:fef4:2856/64 scope link
       valid_lft forever preferred_lft forever
```

now the address for `enp0s20f0` is a ipv4_address for the instance and the SRV record is setup to resolve to this ip address.
the setup-etcd-environment should be able to discover that `147.75.69.215` is the address that should be used to identify this etcd-member.

The change from single interface does increase the complexity of how we discover `ourselves`, but as the number of systems' unicast address is proabably going to be a very small number, this should not add a lot more new cycles for computation.

[1]: https://golang.org/pkg/net/#InterfaceAddrs

**- How to verify it**

**- Description for the changelog**

`allow discovery of etcd-members when using non eth0 as primary interface`

/cc @hexfusion 
/cc @russellb 
kni is also affected by the non-eth0 interfaces.

### CODE REVIEW

**Comment by runcom** [comment]
CI failures seem unrelated. Change LGTM though, I'd love @hexfusion to have a look at this as well

/approve

**Comment by abhinavdahiya** [comment]
Using this patch to create a cluster on Packet.net
```console
[core@adahiya-0-dc6xh-master-0 ~]$ sudo crictl logs -f c9331bd072136
I0313 20:46:20.777540       1 run.go:46] Version: 4.0.0-alpha.0-30-g27973e04-dirty
I0313 20:46:20.821489       1 run.go:132] checking against etcd-0.adahiya-0.devcluster.openshift.com.
I0313 20:46:20.822578       1 run.go:77] dns name is etcd-0.adahiya-0.devcluster.openshift.com
[core@adahiya-0-dc6xh-master-0 ~]$ ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: enp0s20f0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 0c:c4:7a:b5:86:60 brd ff:ff:ff:ff:ff:ff
    inet 147.75.68.155/31 brd 147.75.68.155 scope global dynamic enp0s20f0
       valid_lft 1739sec preferred_lft 1739sec
    inet6 fe80::ec4:7aff:feb5:8660/64 scope link noprefixroute
       valid_lft forever preferred_lft forever
3: enp0s20f1: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq state DOWN group default qlen 1000
    link/ether 0c:c4:7a:b5:86:61 brd ff:ff:ff:ff:ff:ff
[core@adahiya-0-dc6xh-master-0 ~]$ sudo crictl ps -a
CONTAINER ID        IMAGE                                                                                                                         CREATED             STATE               NAME                ATTEMPT             POD ID
81e68a8fecc6a       registry.svc.ci.openshift.org/ci-op-zvg6y6gz/stable@sha256:0f24f7cfcb6c33ca702559940ecf2ee3f01d6511a9a548806564d2820859562d   31 seconds ago      Running             etcd-member         0                   e9c647eb239e8
30215f8a10cf9       quay.io/coreos/kube-client-agent@sha256:d68f85b5ca3adccdc2f4a4c5263f1792798ed44a9b1d63a96004b6e283dc338d                      40 seconds ago      Exited              certs               0                   e9c647eb239e8
c9331bd072136       registry.svc.ci.openshift.org/ci-op-zvg6y6gz/stable@sha256:8883598c0faf0d321e1000a64e8fdae744f12d96aa22a2b4c32e09458ecf9a01   44 seconds ago      Exited              discovery           0                   e9c647eb239e8
```
etcd finds itself from a non-eth0 interface.

File: cmd/setup-etcd-environment/run.go [hunk 1]
```diff
@@ -50,24 +49,28 @@ func runRunCmd(cmd *cobra.Command, args []string) error {
...
-		if err != nil {
-			glog.Errorf("error looking up self: %v", err)
-			return false, nil
-		}
-		if found != "" {
-			dns = found
-			return true, nil
+	var ip string
+	if err := wait.PollImmediate(30*time.Second, 5*time.Minute, func() (bool, error) {
+		for _, cand := range ips {
```
Comment (hexfusion) [inline cmd/setup-etcd-environment/run.go]: This looks good just so I understand why did we reduce the interval here to 30s?

Comment (abhinavdahiya) [reply cmd/setup-etcd-environment/run.go]: the discovery is in the hot path of the bootstrap, I want us to be conservative as dns propagation is slow... and 1 minute is a little long for a poll.

**Comment by kikisdeliveryservice** [comment]
for whatever reason @hexfusion s LGTM didn't go through (bc there were other words on the line i think?), so I will do it again on his behalf :smile: 

/lgtm

**Pull request merged**